core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -361,6 +361,15 @@ public TopDocs call() throws Exception {
                 }
             }
 
+            if (collector != null) {
+                final Collector child = collector;
+                collector = new CancellableCollector(searchContext, collector);
```

Comment:
I'm not a fan of the name (but I don't have better ideas) but this looks good to me.